File: OOP_code/oop.cpp
#include <bits/stdc++.h>
using namespace std;

// Base class
class Animal
{
private:
    string name;

public:
    // Constructor
    Animal(string n) : name(n) {}

    // Getter method
    string getName()
    {
        return name;
    }

    // Virtual method for polymorphism
    virtual void makeSound()
    {
        cout << "Some generic animal sound" << endl;
    }
};

// Derived class
class Dog : public Animal
{
public:
    // Constructor
    Dog(string n) : Animal(n) {}

    // Overriding makeSound method
    void makeSound() override
    {
        cout << getName() << " says: Woof!" << endl;
    }
};

// Another derived class
class Cat : public Animal
{
public:
    // Constructor
    Cat(string n) : Animal(n) {}

    // Overriding makeSound method
    void makeSound() override
    {
        cout << getName() << " says: Meow!" << endl;
    }
};

// Main function
int main()
{
    // Creating objects
    Dog dog("Buddy");
    Cat cat("Whiskers");

    // Encapsulation: Accessing private data through public methods
    cout << "Dog's name: " << dog.getName() << endl;
    cout << "Cat's name: " << cat.getName() << endl;

    // Polymorphism: Different behavior of makeSound for Dog and Cat
    Animal *animals[2];
    animals[0] = &dog;
    animals[1] = &cat;

    for (int i = 0; i < 2; ++i)
    {
        animals[i]->makeSound();
    }

    return 0;
}

File: OOP_code/oop.py
# Class and Object

class Animal:
    def __init__(self, name, species):
        self.name = name
        self.species = species

    def speak(self):
        raise NotImplementedError("Subclasses must implement this method")

# Creating an object of the Animal class
dog = Animal("Buddy", "Dog")
print(dog.name)  # Output: Buddy
print(dog.species) 


# Methods

class Animal:
    def __init__(self, name, species):
        self.name = name
        self.species = species

    def speak(self):
        print(f"{self.name} makes a sound")

# Using the method
dog = Animal("Buddy", "Dog")
dog.speak() 


# Inheritance

class Dog(Animal):
    def speak(self):
        print(f"{self.name} barks")

# Creating an object of the Dog class
dog = Dog("Buddy", "Dog")
dog.speak() 


# Encapsulation


class Animal:
    def __init__(self, name, species):
        self.__name = name  # Private attribute
        self.species = species

    def get_name(self):
        return self.__name

    def set_name(self, name):
        self.__name = name

# Accessing private attribute via public methods
dog = Animal("Buddy", "Dog")
print(dog.get_name())  # Output: Buddy
dog.set_name("Max")
print(dog.get_name()) 


# Polymorphism

class Cat(Animal):
    def speak(self):
        print(f"{self.name} meows")

# Demonstrating polymorphism
animals = [Dog("Buddy", "Dog"), Cat("Whiskers", "Cat")]

for animal in animals:
    animal.speak()
# Output:
# Buddy barks
# Whiskers meows


#  Abstraction

from abc import ABC, abstractmethod

class Animal(ABC):
    def __init__(self, name, species):
        self.name = name
        self.species = species

    @abstractmethod
    def speak(self):
        pass

class Dog(Animal):
    def speak(self):
        print(f"{self.name} barks")

class Cat(Animal):
    def speak(self):
        print(f"{self.name} meows")

# Creating objects of the derived classes
dog = Dog("Buddy", "Dog")
cat = Cat("Whiskers", "Cat")

dog.speak()  # Output: Buddy barks
cat.speak()  

# 